trackintel/preprocessing/positionfixes.py

```diff
@@ -60,114 +66,54 @@ def generate_staypoints(positionfixes, method='sliding',
-
-            # add matching to original positionfixes (for every user)
-
-            for staypoints_id, posfix_idlist in posfix_staypoint_matching.items():
-                # note that we use .loc because above we have saved the id 
-                # of the positionfixes not thier absolut position
-                ret_pfs.loc[posfix_idlist, 'staypoint_id'] = staypoints_id
-
-
+        ret_spts = ret_pfs.groupby('user_id', as_index=False).apply(_generate_staypoints_sliding_user, 
+                                                                    name_geocol, 
+                                                                    elevation_flag,
+                                                                    dist_threshold, 
+                                                                    time_threshold, 
+                                                                    dist_func).reset_index(drop=True)
+        # spts id management
+        ret_spts['id'] = np.arange(len(ret_spts))
+        ret_spts.set_index('id', inplace=True)
+        
+        # Assign staypoint_id to ret_pfs if spts is detected 
+        if not ret_spts.empty:
+            stps2pfs_map = ret_spts[['pfs_id']].to_dict()['pfs_id']
+            
+            ls = []
+            for key, values in stps2pfs_map.items():
+                for value in values:
+                    ls.append([value, key])
+            temp = pd.DataFrame(ls, columns=['id', 'staypoint_id']).set_index('id')
+            ret_pfs = ret_pfs.join(temp, how='left')
+            ret_spts.drop(columns={'pfs_id'}, inplace=True)
+        # if no staypoint is identified
+        else: 
+            ret_pfs['staypoint_id'] = np.NaN
+    
+    # TODO: create tests for dbscan method
+    # TODO: currently only support haversine distance, provode support for other distances, 
+    # we could use the same structure as generate_location function
     elif method == 'dbscan':
         # TODO: Make sure time information is included in the clustering!
         # time information is in the column 'started at', however the user should be able to
         # adjust the distance metric e.g. chebychev
 
-        db = DBSCAN(eps=epsilon, min_samples=num_samples)
-        for user_id_this in ret_pfs['user_id'].unique():
-
-            user_positionfixes = ret_pfs[ret_pfs['user_id'] == user_id_this]  # this is not a copy!
-
-            # TODO: enable transformations to temporary (metric) system
-            transform_crs = None
-            if transform_crs is not None:
-                pass
-
-            # get staypoint matching
-            coordinates = np.array([[g.x, g.y] for g in user_positionfixes['geom']])
-            labels = db.fit_predict(coordinates)
-
-            # add positionfixes - staypoint matching to original positionfixes
-            ret_pfs.loc[user_positionfixes.index, 'staypoint_id'] = labels
-
+        ret_pfs = ret_pfs.groupby("user_id").apply(_generate_staypoints_dbscan_user, 
```

Comment:
I didn't try it out but doesn't this generate non-unique staypoint ids? E.g. the staypoint ID starts at 0 for every user?

Reply:
Yes! you caught me 💯. The original function has also this behaviour and this is left as a todo (also because there are no tests )... We can also use the code in `generate_location`, they are really similar.

Reply:
Ah ok. There I didn't notice iin the places case ;-)
Can you document this/these problem/s somehow before we forget it. I think it would be enough to add a #todo note in the code and open an issue for it.